src/subscription/subscribe.js

```diff
@@ -0,0 +1,184 @@
+  // AsyncIterable yielding raw payloads.
+  const subscription = resolveSubscription(
+    exeContext,
+    exeContext.operation,
+    rootValue
+  );
+
+  // For each payload yielded from a subscription, map it over the normal
+  // GraphQL `execute` function, with `payload` as the rootValue.
+  return mapAsyncIterator(
+    subscription,
+    payload => execute(
+      schema,
+      document,
+      payload,
+      contextValue,
+      variableValues,
+      operationName
+    )
+  );
+}
+
+function resolveSubscription(
+  exeContext: ExecutionContext,
+  operation: OperationDefinitionNode,
+  rootValue: mixed
+): AsyncIterator<mixed> {
+  // Note: this function is almost the same as executeOperation() and
+  // resolveField() with only a few minor differences.
+
+  const type = getOperationRootType(exeContext.schema, exeContext.operation);
+  const fields = collectFields(
+    exeContext,
+    type,
+    exeContext.operation.selectionSet,
+    Object.create(null),
+    Object.create(null)
+  );
+
+  const responseNames = Object.keys(fields);
+  invariant(
+    responseNames.length === 1,
+    'A subscription must contain exactly one field.'
+  );
+  const responseName = responseNames[0];
+  const fieldNodes = fields[responseName];
+  const fieldPath = addPath(undefined, responseName);
+
+  const fieldNode = fieldNodes[0];
+  const fieldName = fieldNode.name.value;
+  const fieldDef = getFieldDef(exeContext.schema, type, fieldName);
+  invariant(
+    fieldDef,
+    'This subscription is not defined by the schema.'
+  );
+
+  // TODO: make GraphQLSubscription flow type special to support defining these?
+  const resolveFn = (fieldDef: any).subscribe || defaultFieldResolver;
+
+  // The resolve function's optional third argument is a context value that
+  // is provided to every resolve function within an execution. It is commonly
+  // used to represent an authenticated user, or request-specific caches.
+  const context = exeContext.contextValue;
+
+  // The resolve function's optional fourth argument is a collection of
+  // information about the current execution state.
+  const info: GraphQLResolveInfo = {
+    fieldName,
+    fieldNodes,
+    returnType: fieldDef.type,
+    parentType: type,
+    path: fieldPath,
+    schema: exeContext.schema,
+    fragments: exeContext.fragments,
+    rootValue: exeContext.rootValue,
+    operation: exeContext.operation,
+    variableValues: exeContext.variableValues,
+  };
+
+  // Build a JS object of arguments from the field.arguments AST, using the
+  // variables scope to fulfill any variable references.
+  const args = getArgumentValues(
+    fieldDef,
+    fieldNode,
+    exeContext.variableValues
+  );
+
+  // TODO: resolveFn could throw!
+  const subscription = resolveFn(rootValue, args, context, info);
+
+  invariant(
+    isIterable(subscription),
+    'Subscription must return async-iterator.'
+  );
+
+  return subscription;
+}
```

Comment:
The large majority of this function is recycled from `executeOperation()` and `resolveField()`. There's almost nothing new here.

The differences: 

* It's only resolving a single field instead of an object of fields.
* It will call `subscribe` instead of calling `resolve` if defined on the schema
* A couple extra invariants